File: leetcode/simplify_path.py
# Given an absolute path for a file (Unix-style), simplify it.
# Time Complexity: O(n)
# Space Complexity: O(n)

class Solution:
    # @param {string} path
    # @return {string}
    def simplifyPath(self, path):
        result = []
        for token in (str for str in path.split("/") if str):
            if token == "..":
                if len(result) > 0: result.pop()
            elif token != ".":
                result.append(token)
        return '/' + '/'.join(result)

if __name__ == "__main__":
    solution = Solution()
    print solution.simplifyPath("/home/")
    print solution.simplifyPath("/a/./b/../../c/")

File: leetcode/simplify_path.js
/* Given an absolute path for a file (Unix-style), simplify it.
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {string} path
 * @return {string}
 */
var simplifyPath = function(path) {
    var result = []
    for (var str of path.split("/")) {
        if (str.length === 0)
            continue;
        if (str === "..") {
            if (result.length > 0) {
                result.pop();
            }
        } else if (str !== ".") {
            result.push(str);
        }
    }
    return "/".concat(result.join("/"));
};
